streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java

```diff
@@ -197,31 +197,33 @@ public String newStoreName(final String prefix) {
         return prefix + String.format(KTableImpl.STATE_STORE_NAME + "%010d", index.getAndIncrement());
     }
 
-    public synchronized void addStateStore(final StoreBuilder builder) {
+    public synchronized void addStateStore(final StoreBuilder<?> builder) {
         addGraphNode(root, new StateStoreNode(builder));
     }
 
-    public synchronized void addGlobalStore(final StoreBuilder<KeyValueStore> storeBuilder,
-                                            final String sourceName,
-                                            final String topic,
-                                            final ConsumedInternal consumed,
-                                            final String processorName,
-                                            final ProcessorSupplier stateUpdateSupplier) {
-
-        final StreamsGraphNode globalStoreNode = new GlobalStoreNode(storeBuilder,
-                                                                     sourceName,
-                                                                     topic,
-                                                                     consumed,
-                                                                     processorName,
-                                                                     stateUpdateSupplier);
+    public synchronized <K, V> void addGlobalStore(final StoreBuilder<KeyValueStore<K, V>> storeBuilder,
+                                                   final String sourceName,
+                                                   final String topic,
+                                                   final ConsumedInternal<K, V> consumed,
+                                                   final String processorName,
+                                                   final ProcessorSupplier<K, V> stateUpdateSupplier) {
+
+        final StreamsGraphNode globalStoreNode = new GlobalStoreNode<>(
+                storeBuilder,
+                sourceName,
+                topic,
+                consumed,
+                processorName,
+                stateUpdateSupplier
+        );
 
         addGraphNode(root, globalStoreNode);
     }
 
-    public synchronized void addGlobalStore(final StoreBuilder<KeyValueStore> storeBuilder,
-                                            final String topic,
-                                            final ConsumedInternal consumed,
-                                            final ProcessorSupplier stateUpdateSupplier) {
+    public synchronized <K, V> void addGlobalStore(final StoreBuilder<KeyValueStore<K, V>> storeBuilder,
```

Comment:
Ditto here: are we always restricting to a kv-store.

Reply:
Not that the restriction is applied in the existing code already! (old L221 above).

However, I just did a quick check and the specified type from above does actually not result in a class cast exception, ie, the compiler does not insert a cast and we can also have global WindowStores (for example). Hence, I am going to remove the store type to not break compatiblitily.